draft-reddy-ipsecme-ikev2-pqc-auth.md

```diff
@@ -131,6 +132,53 @@ SLH-DSA does not introduce a new hardness assumption beyond those inherent to th
+
+The first is to note that both ML-DSA and SLH-DSA have prehashed parameter sets; that is, ones designed to sign a message that has been hashed by an external source.
+At first place, this would appear to be an ideal solution, however it turns out that there are a number of practical issues.
+The first is that the prehashed version of ML-DSA and SLH-DSA would appear to be rarely used, and so it is not unlikely that support for it within crypto libraries may be lacking.
+The second is that the public keys for the prehashed versions use different OIDs; this means that the certificates for IKEv2 would necessarily be different than certificates for other protocols (and some CAs might not support issuing certificates for prehashed ML-DSA or prehashed SLH-DSA, again because of the lack of use).
+The third is that some users have expressed a desire not to use the prehashed parameter sets.
+
+The second is to note that, while IKEv2 normally acts this way, it doesn't always.
+EdDSA has a similar constraint on not working cleanly with the standard 'hash and then sign' paradigm, and so the existing [RFC8420] provides an alternative method, which ML-DSA would cleanly fit into.
+We could certainly adopt this same strategy; our concern would be that it might be more difficult for IKEv2 implementors which do not already have support for EdDSA.
+
+The third way is what we can refer to as 'fake prehashing'; IKEv2 would generate the hash as current, but instead of running ML-DSA or SLH-DSA in prehash mode, we have itsign it in pure mode as if it was the message.
+This is a violation of the spirit, if not the letter of FIPS 204, 205
+However, it is secure (assuming the hash function is strong), and fits in cleanly with both the existing IKEv2 architecture, and what crypto libraries provide.
+On the other hand, for SLH-DSA, this means that we're now dependent on collision resistance (while the rest of the SLH-DSA architecture was carefully designed not to be).
+
+## Context String
+
+An additional feature that ML-DSA, SLH-DSA provide it allows the signer and the verifier to provide a 'context string'.
+The signature would verify only if the context strings that are provided by both the signer and the verifier match.
+The reason behind this is to ensure that if a public key is used for multiple purposes, a signature for one purpose cannot be used by an adversary in another.
+In our case, if the same certificate where used to sign both IKEv2 and TLS exchanges, an adversary could not possibly take the signature from an TLS exchange and try to use it within an IKEv2 exchange.
+
+This really is not a necessary security safe guard; the messages that are actually signed in both cases are distinct enough that an adversary could not actually take advantage of this, even without the protection.
+
+However, given that ML-DSA and SLH-DSA do provide such a service, and it appears that crypto libraries do support a nonempty context, we cannot see a reason not to use it.
```

Comment:
It is not clear to me what the context string should be set to for interoperability purpose ?
I don't see it set in any of TLS, JOSE and LAMPS WG drafts.

Reply:
As for what it should be set to, in this draft the IKE signer and the IKE verifier set it to the same thing (and the whole point is that those signatures are not interoperable with other use)
As for LAMPS, the current composite signature draft does also set it.
As far as I can see, the only reason not to do it is if there are ML-DSA, SLH-DSA implementations that do not support it.  Currently, that is unknown (the ones that I've checked with do)

Reply:
My concern here is that in this case signing with ML-DSA/SLH-DSA should be handled differently by IKEv2 than signing with other algorithms. Currently the set of input for the crypto engine that performs signing on behalf of IKE doesn't depend on the signature algorithm. If the context string have to be provided, then IKE implementations have to provide it only if ML-DSA/SLH-DSA is used. Thus, the input data for signing will have to be depend on the algorithm. Same with verifying.